File: sdk/javascript/src/exceptions.ts
/**
 * Exception classes for the AI Trading Platform SDK
 */

export class TradingPlatformError extends Error {
  public statusCode?: number;
  public responseData?: Record<string, any>;

  constructor(message: string, statusCode?: number, responseData?: Record<string, any>) {
    super(message);
    this.name = 'TradingPlatformError';
    this.statusCode = statusCode;
    this.responseData = responseData;
  }
}

export class AuthenticationError extends TradingPlatformError {
  constructor(message: string, statusCode?: number, responseData?: Record<string, any>) {
    super(message, statusCode, responseData);
    this.name = 'AuthenticationError';
  }
}

export class AuthorizationError extends TradingPlatformError {
  constructor(message: string, statusCode?: number, responseData?: Record<string, any>) {
    super(message, statusCode, responseData);
    this.name = 'AuthorizationError';
  }
}

export class RateLimitError extends TradingPlatformError {
  public retryAfter?: number;

  constructor(message: string, retryAfter?: number, statusCode?: number, responseData?: Record<string, any>) {
    super(message, statusCode, responseData);
    this.name = 'RateLimitError';
    this.retryAfter = retryAfter;
  }
}

export class ValidationError extends TradingPlatformError {
  public validationErrors: any[];

  constructor(message: string, validationErrors: any[] = [], statusCode?: number, responseData?: Record<string, any>) {
    super(message, statusCode, responseData);
    this.name = 'ValidationError';
    this.validationErrors = validationErrors;
  }
}

export class APIError extends TradingPlatformError {
  constructor(message: string, statusCode?: number, responseData?: Record<string, any>) {
    super(message, statusCode, responseData);
    this.name = 'APIError';
  }
}

export class NetworkError extends TradingPlatformError {
  constructor(message: string, statusCode?: number, responseData?: Record<string, any>) {
    super(message, statusCode, responseData);
    this.name = 'NetworkError';
  }
}

export class TimeoutError extends TradingPlatformError {
  constructor(message: string, statusCode?: number, responseData?: Record<string, any>) {
    super(message, statusCode, responseData);
    this.name = 'TimeoutError';
  }
}

export class WebSocketError extends TradingPlatformError {
  constructor(message: string, statusCode?: number, responseData?: Record<string, any>) {
    super(message, statusCode, responseData);
    this.name = 'WebSocketError';
  }
}

export class PluginError extends TradingPlatformError {
  constructor(message: string, statusCode?: number, responseData?: Record<string, any>) {
    super(message, statusCode, responseData);
    this.name = 'PluginError';
  }
}

File: sdk/python/ai_trading_platform/exceptions.py
"""
Exception classes for the AI Trading Platform SDK
"""

from typing import Optional, Dict, Any


class TradingPlatformError(Exception):
    """Base exception for all SDK errors"""
    
    def __init__(self, message: str, status_code: Optional[int] = None, 
                 response_data: Optional[Dict[str, Any]] = None):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.response_data = response_data or {}


class AuthenticationError(TradingPlatformError):
    """Raised when authentication fails"""
    pass


class AuthorizationError(TradingPlatformError):
    """Raised when user lacks permission for requested action"""
    pass


class RateLimitError(TradingPlatformError):
    """Raised when API rate limit is exceeded"""
    
    def __init__(self, message: str, retry_after: Optional[int] = None, **kwargs):
        super().__init__(message, **kwargs)
        self.retry_after = retry_after


class ValidationError(TradingPlatformError):
    """Raised when request validation fails"""
    
    def __init__(self, message: str, validation_errors: Optional[list] = None, **kwargs):
        super().__init__(message, **kwargs)
        self.validation_errors = validation_errors or []


class APIError(TradingPlatformError):
    """Raised for general API errors"""
    pass


class NetworkError(TradingPlatformError):
    """Raised for network-related errors"""
    pass


class TimeoutError(TradingPlatformError):
    """Raised when request times out"""
    pass


class WebSocketError(TradingPlatformError):
    """Raised for WebSocket-related errors"""
    pass


class PluginError(TradingPlatformError):
    """Raised for plugin-related errors"""
    pass